### PR TITLE
New Rule: prefer-for-of

### PR DESCRIPTION
**Pull request by chrismbarr**
Based on discussions in #1335, i've made a new rule: `prefer-for-of`! 

**Reasoning**
Basically a standard `for` loop should only be allowed if the incrementer variable is used in the block of the loop for **anything other than accessing the current array item**.

 👍 **This is Fine** since the array index is needed for something
```ts
for (var j = 0; j <= arr.length; j++) {
    doMath(j);
}
```

👎 **This could be replaced with a `for(...of...)` loop** since the index is only used to access the array
```ts
for (var i = 0; i <= arr.length; i++) {
    doThing(arr[i]);
}
```

It will fail with the text:
>This standard for loop could be replaced with a for(... of ...) loop

👍 **Here's the preferred way to write the above loop**
```ts
for (var item of arr) {
    doThing(item);
}
```

This will also catch for loops that use syntax like: `for(var t=0, len=arr.length; t < len; t++)`

### CODE REVIEW

**Comment by JKillian** [comment]
Sorry for the delay getting to this and other PRs! Not sure when I will, but promise I will eventually!

File: README.md [hunk 1]
```diff
@@ -306,6 +306,7 @@ Core rules are included in the `tslint` package.
 * `one-variable-per-declaration` disallows multiple variable definitions in the same statement.
   * `"ignore-for-loop"` allows multiple variable definitions in for loop statement.
 * `only-arrow-functions` disallows traditional `function () { ... }` declarations, preferring `() => { ... }` arrow lambdas.
+* `prefer-for-of` recommends a [TypeScript `for(... of ...)` loop](https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html) over a standard `for` loop if the index is only used to access the array being looped over.
```
Comment (JoshuaKGoldberg) [inline README.md]: This is an ES2015 construct, no?

Comment (chrismbarr) [inline README.md]: Well so it is!  I was not aware, I thought it was just TypeScript syntactic sugar.

**Comment by JKillian** [comment]
Can you merge develop into your branch so the tests will pass? Will review this PR this week!

**Comment by chrismbarr** [comment]
I was on vacation last week - back now!  I'll try and get my PR's fixed.

ugh.... this master merge really got screwed up.  Right now it might be easier for me to just create a new PR with a single commit...

**Comment by chrismbarr** [comment]
closing this since there are so many commits and merge issues with master due to the age of this branch.  See #1435 instead.

**Pull request closed without merging**